src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/CustomAttributeWrapper.cs

```diff
@@ -0,0 +1,173 @@
+        private readonly byte[] _binaryAttribute;
+
+        public CustomAttributeWrapper(ConstructorInfo constructorInfo, ReadOnlySpan<byte> binaryAttribute)
+        {
+            _constructorInfo = constructorInfo;
+            _binaryAttribute = binaryAttribute.ToArray(); // TODO: Update to BlobHandle when public API public APi for MetadataBuilder.GetOrAddBlob(ReadOnlySpan<byte>) added
+        }
+
+        public ConstructorInfo Ctor => _constructorInfo;
+        public byte[] Data => _binaryAttribute;
+    }
+
+    internal struct CustomAttributeInfo
+    {
+        public ConstructorInfo _ctor;
+        public object?[] _ctorArgs;
+        public string[] _namedParamNames;
+        public object?[] _namedParamValues;
+        private const int Field = 0x53;
+        private const int EnumType = 0x55;
+        private const int NullValue = 0xff;
+        private const int OneByteMask = 0x7f;
+        private const int TwoByteMask = 0x3f;
+        private const int FourByteMask = 0x1f;
+
+        internal static CustomAttributeInfo DecodeCustomAttribute(ConstructorInfo ctor, ReadOnlySpan<byte> data)
+        {
+            int pos = 2;
+            CustomAttributeInfo info = default;
+
+            if (data.Length < 2)
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidCustomAttributeLength, ctor.DeclaringType, data.Length));
+            }
+            if ((data[0] != 0x01) || (data[1] != 0x00))
+            {
+                throw new InvalidOperationException(SR.Format(SR.InvalidOperation_InvalidProlog, ctor.DeclaringType));
+            }
+
+            ParameterInfo[] pi = ctor.GetParameters();
+            info._ctor = ctor;
+            info._ctorArgs = new object?[pi.Length];
+            for (int i = 0; i < pi.Length; ++i)
+            {
+                info._ctorArgs[i] = DecodeCustomAttributeValue(pi[i].ParameterType, data, pos, out pos);
+            }
+            int numNamed = data[pos] + (data[pos + 1] * 256);
+            pos += 2;
+
+            info._namedParamNames = new string[numNamed];
+            info._namedParamValues = new object[numNamed];
+            for (int i = 0; i < numNamed; ++i)
+            {
+                int namedType = data[pos++];
+                int dataType = data[pos++];
+
+                if (dataType == EnumType)
+                {
+                    // skip bytes for Enum type name;
+                    int len2 = DecodeLen(data, pos, out pos);
+                    pos += len2;
+                }
+
+                int len = DecodeLen(data, pos, out pos);
+                string name = StringFromBytes(data, pos, len);
+                info._namedParamNames[i] = name;
+                pos += len;
+
+                if (namedType == Field)
+                {
+                    Type fieldType = dataType == EnumType ? typeof(int) : ElementTypeToType((PrimitiveSerializationTypeCode)dataType);
```

Comment:
`typeof(int)` is ok here only because this is used for set of known CAs and all enums in these CAs are int32.

It may be worth a comment.